Add doc for task continuous deployment (#246)
spring-io/dataflow.spring.io: content/documentation/pages/4-batch-developer-guides/4-continuous-deployment/1-cd-basics.md

---
path: 'batch-developer-guides/continuous-deployment/cd-basics/'
title: 'Continuous Deployment of task applications'

> description: 'Continuous Deployment of Task applications'

Reviewer: Update the description to be something different from the title.   i.e. This section will discuss how to use Continuous Deployment of Tasks in SCDF.
If you agree I can update on merge.

Author: Yeah, that is better. Please go ahead!

---

# Continuous Deployment of task applications

As task applications evolve, you want to get your updates to production.
The change can be a new version of the task application that fixes a bug or setting a deployment property different from the previous task launch.

> When a task application is registered, a version is associated with it. A task application can have multiple versions associated with it, with one selected as the default. The following image illustrates an application with multiple versions associated with it (see the timestamp entry).

Reviewer 2: ```suggestion
When a task application is registered in SCDF, a version is typically associated with it. A task application can have multiple versions associated with it, with one selected as the default. The following image illustrates an application with multiple versions associated with it (see the timestamp entry).
```

<img src="images/scdf-task-application-versions.png" alt="Registering Task Applications with multiple Versions" width="740"/>

> Versions of an application are managed by registering multiple applications with the same name and coordinates, _except_ the version. For example, if you were to register an application with the following values, you would get one application registered with two versions (2.0.0.RELEASE and 2.1.0.RELEASE):

Reviewer 2: ```suggestion
Versions of an application are managed in SCDF by registering multiple applications with the same name and coordinates, _except_ the version will be different. For example, if you were to register an application with the following values, you would get one application registered with two versions (2.0.0.RELEASE and 2.1.0.RELEASE):
```

- Application 1
  ** Name: `timestamp`
  ** Type: `task`
  \*\* URI: `maven://org.springframework.cloud.task.app:timestamp-task:2.0.0.RELEASE`
- Application 2
  ** Name: `timestamp`
  ** Type: `task`
  \*\* URI: `maven://org.springframework.cloud.task.app:timestamp-task:2.1.0.RELEASE`

Besides having multiple versions, Spring Cloud Data Flow needs to know which version to run on the next launch. This is indicated by setting a version to be the default version. Whatever version of a task application is configured as the default version is the one to be run on the next launch request. You can see which version is the default in the UI as this image shows:

<img src="images/scdf-task-default-version.png" alt="Task Application Default Version" width="740"/>

## Task Launch Lifecycle

> In previous versions of Spring Cloud Data Flow, when the request to launch a task was received, Spring Cloud Data Flow would deploy the application (if needed) and run it. If the application was being run on a platform that did not need to have the application deployed every time (CloudFoundry for example), the previously deployed application was used. This flow has changed in 2.3. The following image shows what happens when a task launch request comes in now:

Reviewer 2: ```suggestion
Before the CD support for Tasks in SCDF, when the request to launch a task was received, Spring Cloud Data Flow would deploy the application (if needed) and run it. If the application was being run on a platform that did not need to have the application deployed every time (CloudFoundry for example), the previously deployed application was used. This flow has changed starting from 2.3. The following image shows what happens when a task launch request comes in now:
```
